src/coreclr/jit/lower.cpp

```diff
@@ -3369,9 +3295,63 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
     }
 #endif // TARGET_ARM64 || TARGET_LOONGARCH64
 
-    ContainCheckJTrue(jtrue);
-
+    assert(relop->OperIsCompare());
+    assert(relop->gtNext == jtrue);
     assert(jtrue->gtNext == nullptr);
+
+    GenCondition cond = GenCondition::FromRelop(relop);
+#ifdef TARGET_XARCH
+    // Optimize FP x != x to only check parity flag. This is a common way of
+    // checking NaN and avoids two branches that we would otherwise emit.
+    if ((cond.GetCode() == GenCondition::FNEU) && relopOp1->OperIsLocal() && GenTree::Compare(relopOp1, relopOp2) &&
+        IsInvariantInRange(relopOp1, relop) && IsInvariantInRange(relopOp2, relop))
+    {
+        cond = GenCondition(GenCondition::P);
+    }
+#endif
+
+    // Optimize EQ/NE(op_that_sets_zf, 0) into op_that_sets_zf with GTF_SET_FLAGS.
+    if (relop->OperIs(GT_EQ, GT_NE) && relopOp2->IsIntegralConst(0) && relopOp1->SupportsSettingZeroFlag() &&
+        IsInvariantInRange(relopOp1, relop))
+    {
+        relopOp1->gtFlags |= GTF_SET_FLAGS;
+        relopOp1->SetUnusedValue();
+
+        BlockRange().Remove(relopOp1);
+        BlockRange().InsertBefore(jtrue, relopOp1);
+        BlockRange().Remove(relop);
+        BlockRange().Remove(relopOp2);
+    }
+    else
+    {
+        relop->gtType = TYP_VOID;
+        relop->gtFlags |= GTF_SET_FLAGS;
+
+        if (relop->OperIs(GT_EQ, GT_NE, GT_LT, GT_LE, GT_GE, GT_GT))
```

Comment:
With this code, a `JTRUE(LT)` is always turned into a `JCC(CMP)` with the `JCC` containing a condition code within gtCondition. In codegen for `JCC`, we don't need to check the child for the condition. I'm happy with that.

In my and chains patch, I let optimise bools create a chain of the form `NE(AND(LT,EQ),0)`. LowerConstCompare() turns it into `TEST_NE(LT,EQ)` and then new function `lowerTestCompare()` turns it into `CCMP_LT(EQ)`. A `JTRUE` node is allowed to have either a standard compare child or a CCMP child, eg `JTRUE(CCMP_LT(EQ)`.

When combing both worlds, the lowering stage gets a little complicated, but everything else feels simple:

The `CCMP_LT` has a gtCondition, plus there is also a `CCMP` node. `lowerTestCompare()` turns `TEST_NE(LT,EQ)` into `CCMP_LT(CMP)`. Then `LowerJTrue()` would turn it into `CCMP(CMP)`, giving us `JCC(CCMP(CMP)`.

Reply:
Right. After this PR the plan is that we then introduce `GT_SELECTCC` which is to `GT_SELECT` as `GT_JCC` is to `GT_JTRUE` (with the additional caveat that we never see `GT_JTRUE` in codegen, because we have an invariant that there is nothing between the relop and `GT_JTRUE`).

When lowering `GT_SELECT` and `GT_JTRUE` we will invoke a common function to transform compare chains into `CMP + CCMP + ... `, followed up by the transformation into `GT_SELECTCC` and `GT_JCC`. For `GT_SELECT` this function also needs to verify that reordering the compare to happen before `GT_SELECT` is safe (if it isn't we'll leave `GT_SELECT` as `GT_SELECT` instead of transforming into `GT_SELECTCC`). This should put everything on the same plan and I think it will unify the last of the handling.

Reply:
FWIW, it sounds like `CCMP_LT` would not be necessary with generic `CCMP`. I.e. we can transform `TEST_NE(LT,EQ)` into `SETCC(CCMP(CMP(x, y)))` and teach `LowerJTrue` to pass through `SETCC` nodes.
I actually considered using this approach instead of having `GT_TEST_NE/GT_TEST_EQ/GT_BITTEST_NE/GT_BITTEST_EQ` but it creates an additional node and will require some bespoke interference logic for `GT_SELECT` (but I think that logic will be necessary on arm64 in that case regardless).
I can try to see what it would look like in a follow-up.

Reply:
> we can transform `TEST_NE(LT,EQ)` into `SETCC(CCMP(CMP(x, y)))`

That's better. Reduces the number of node types.

> When lowering GT_SELECT and GT_JTRUE we will invoke a common function to transform compare chains into CMP + CCMP + ... 

If the chain parsing is also done at the CMP and AND nodes, then we should be able to only check the parent and child at each stage, so no probably no need to recursively go through the chain.

Reply:
So, my question then is what to do with my and chains patch? I should wait until this patch is merged before continuing with it. Should I also wait for the GT_SELECTCC or any other patches - I don't think I need to, they should be mostly independent, but will merge conflict.

Reply:
I think waiting for this one is a good idea, I hope we can merge it by today or tomorrow. Not sure how much this one will conflict with your PR.
I don't think you need to wait for `GT_SELECTCC` either. I will probably move the xarch backend to `GT_SELECTCC` as the first step and leave arm64 alone, so then they shouldn't conflict.

Reply:
> For `GT_SELECT` this function also needs to verify that reordering the compare to happen before `GT_SELECT` is safe (if it isn't we'll leave `GT_SELECT` as `GT_SELECT` instead of transforming into `GT_SELECTCC`). This should put everything on the same plan and I think it will unify the last of the handling.

This will drop all the containing within `SELECT`? ie -`SELECT` will always load flags from a register, but SELECTCC will require the previous node to be a compare (like JCC) and always rely of the condition flags being set.

Reply:
Yes, exactly.